specs/phase0/beacon-chain.md

```diff
@@ -729,14 +728,18 @@ def compute_shuffled_index(index: uint64, index_count: uint64, seed: Bytes32) ->
 
     # Swap or not (https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf)
     # See the 'generalized domain' algorithm on page 3
-    for current_round in range(SHUFFLE_ROUND_COUNT):
-        pivot = bytes_to_int(hash(seed + int_to_bytes(current_round, length=1))[0:8]) % index_count
+    for current_round in map(uint64, range(SHUFFLE_ROUND_COUNT)):
```

Comment:
This is the type of casting I'm most uneasy with. It just feels like a Python hack.
Does the following work for mypy?
```python
current_round: uint64
for current_round in range(SHUFFLE_ROUND_COUNT):
```
Or is that too verbose for spec also?

Reply:
It doesn't work for mypy since the issue is that the `range(SHUFFLE_ROUND_COUNT)` generator returns `int`.

Reply:
The `current_round` is also only converted to a single byte (see other comment about `uint8`). Avoiding this `map` would be nice.

Reply:
What about changing the type of the first parameter of `int_to_bytes` to `int` ?
`current_round` can be kept `int` then.

Alternatively one can introduce something like `urange` method, which will be the same `map` applied to `range`, though.

```python
def urange(*args):
   return map(uint64, range(*args))
```

As `uint64` is heavily used in pyspec, introducing unsigned `range` looks pretty reasonable.

Reply:
@protolambda 

1. Agreed with that `current_round` can be `uint8`.
2. Currently, `uint8` is not compatible with `uint64` for `mypy` in remerkerble layer when we want to pass it to `int_to_bytes`. Although it's fine to set `current_round` to `uint8` type in pyspec right now since we haven't changed our CI linter setting to as strict as #1707.
3. For implementation-wise, I feel having `int_to_bytes` be an instance function in SSZ level would be neater. 
    - We can define `uint.uint_to_bytes` in `remerkerble`:

    ```python
    def uint_to_bytes(self, byteorder: str) -> bytes:
         return self.to_bytes(self.type_byte_length, byteorder=byteorder)
    ```
    - To decrease the requirements for pyspec readers to understand remerkerble, we can modify `int_to_bytes` to:
    ```python
    def int_to_bytes(n: uint) -> bytes:
        """
        Return the serialization of ``n`` in ``ENDIANNESS``-endian in the byte length of ``uint`` type.
        """
        return uint.uint_to_bytes(byteorder=ENDIANNESS)
    ```

What do you think?

---

@ericsson49 
Hi!
1. To me, the goal is setting a boundary of all variables in pyspec since Python `int` is unbounded. I hope it would decrease the ambiguity and make other implementers' lives better?
2. IMHO It seems better to have `urange` supports all `uint` types. Especially if we want to set `current_round` to `uint8` as @protolambda suggested. I'd say if we really want to avoid `map` for readability, the simplest way is:
```python
for current_round in range(SHUFFLE_ROUND_COUNT):
    current_round = uint8(current_round)  # cast `int` to `uint8` type
```
What do you think?

3. What do you think about the above proposal for `int_to_bytes` change?

---
Thank you both! 🙂

Reply:
> We can define uint.uint_to_bytes in remerkerble:

We already have that, as well as stream variants, for every type :).

But for spec style and less hidden things, we generally avoid it.

The methods:
- `def encode_bytes(self) -> bytes:`
- `def serialize(self, stream: BinaryIO) -> int:`
- `def decode_bytes(cls: Type[V], bytez: bytes) -> V:`
- `def deserialize(cls: Type[V], stream: BinaryIO, scope: int) -> V:`

So `uint8(123).encode_bytes()` would just work to get it as bytes.

And there are a ton more util methods for size bounds, working with merkle backings, etc., but it's meant for tooling more than the spec.

Reply:
@protolambda 

Oh great!

> But for spec style and less hidden things, we generally avoid it.

Agreed with that we should generally avoid exposing SSZ APIs. Another solution is that we can describe `int_to_bytes(n: uint)` in the same way we describe `hash_tree_root`.

Reply:
Created an alternative solution PR: https://github.com/ethereum/eth2.0-specs/pull/1935

Reply:
@hwwhww
Hi! Sorry, I missed a notification, so responding somewhat late.

> To me, the goal is setting a boundary of all variables in pyspec since Python int is unbounded. I hope it would decrease the ambiguity and make other implementers' lives better?

That's a great goal. It definitely helps. Basically, a typical developer would prefer using a bounded integer type whenever possible. And resolving possible range can be a problem.
In the particular case, it's however obvious that `current_round` is bounded.
But as `current_round` is intended to be `uint8` it definitely worth using an explicit type/conversion here.

However, unbounded `int` cannot be completely avoided, I guess, e.g. `legendre_bit` uses it and private keys are sometimes used. But it would be definitely great to separate cases where bounded `int`s are enough from those, where arbitrary precision `int`s are more appropriate.

> IMHO It seems better to have urange supports all uint types. Especially if we want to set current_round to uint8 as @protolambda suggested. I'd say if we really want to avoid map for readability, the simplest way is:

I think it's a matter of taste. Basically, all variants are clear, but some slightly less concise.